cmd/metacache-stream.go

```diff
@@ -817,36 +816,62 @@ func (w *metacacheBlockWriter) Close() error {
 type metacacheBlock struct {
 	data  []byte
 	n     int
-	First string `json:"f"`
-	Last  string `json:"l"`
-	EOS   bool   `json:"eos,omitempty"`
+	first string
+	last  string
+	eos   bool
+}
+
+const metacacheBlockReservedHeaderPrefix = ReservedMetadataPrefixLower + "metacache-part-"
+
+// This code is specifically written in this form to avoid json unmarshalling
+//
+// BenchmarkKVSimple-4     13431099                88.15 ns/op           48 B/op          1 allocs/op
+// v/s
+// BenchmarkKVJSON-4        1510602               789.4 ns/op           224 B/op          4 allocs/op
+//
+// So that only allocation this function ever does is for strings.SplitN()
+func (b *metacacheBlock) loadHeaderKV(meta map[string]string, partKey string) error {
+	v, ok := meta[partKey]
+	if !ok {
+		return io.ErrUnexpectedEOF
+	}
+	parts := strings.SplitN(v, ":", 3)
+	if len(parts) != 3 {
+		return io.ErrUnexpectedEOF
+	}
+	b.first = parts[0]
+	b.last = parts[1]
+	b.eos = parts[2] == "true"
+	return nil
 }
 
 func (b metacacheBlock) headerKV() map[string]string {
-	var json = jsoniter.ConfigCompatibleWithStandardLibrary
-	v, err := json.Marshal(b)
-	if err != nil {
-		logger.LogIf(context.Background(), err) // Unlikely
-		return nil
-	}
-	return map[string]string{fmt.Sprintf("%s-metacache-part-%d", ReservedMetadataPrefixLower, b.n): string(v)}
+	// Code specifically written like this to avoid any allocations
+	// BenchmarkKVSimple-4     29415350                40.29 ns/op            0 B/op          0 allocs/op
+	// v/s
+	// BenchmarkKVJSON-4        3852229               267.3 ns/op            24 B/op          1 allocs/op
+	//
+	// So that only allocation this function ever does is for its returned value.
+	key := metacacheBlockReservedHeaderPrefix + strconv.Itoa(b.n)
+	value := b.first + ":" + b.last + ":" + strconv.FormatBool(b.eos)
```

Comment:
This will break if any prefix or object name contains a `:`. We can control that for erasure, but not for fs.